language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -82,79 +111,150 @@ mergePackageMap ps = foldM merge Map.empty ps
+    sdkVersion <- fromMaybe "0.0.0" <$> getSdkVersionMaybe
+    let damlTypesVersion = fromMaybe sdkVersion optDamlTypesVersion
+       -- The '@daml/types' version that the packages we write should depend on.
+    putStrLn $ "Referencing '@daml/types' : \"" <> damlTypesVersion <> "\""
+    let packageVersion   = sdkVersion
+       -- The version string for packages we write.
     ps <- readPackages optInputDars
     case mergePackageMap ps of
       Left err -> fail . T.unpack $ err
-      Right pm ->
-        forM_ (Map.toList pm) $
-        \(pkgId, (mbPkgName, pkg)) -> do
-          let id = show $ unPackageId pkgId
-              name = maybe id (T.unpack . unPackageName) mbPkgName
-              asName = if name == id then "itself" else name
-          putStrLn $ "Generating " <> id <> " as " <> asName
-          daml2ts opts pm pkgId pkg mbPkgName
-
-daml2ts :: Options -> Map.Map PackageId (Maybe PackageName, Package) -> PackageId -> Package -> Maybe PackageName -> IO ()
-daml2ts Options{..} pm pkgId pkg mbPkgName = do
-    let outputDir = optOutputDir </> (T.unpack . maybe (unPackageId pkgId) unPackageName) mbPkgName
-    createDirectoryIfMissing True outputDir
-    T.writeFileUtf8 (outputDir </> "packageId.ts") $ T.unlines
-        ["export default '" <> unPackageId pkgId <> "';"]
-    forM_ (packageModules pkg) $ \mod -> do
-        whenJust (genModule pm pkgId mod) $ \modTxt -> do
-            let outputFile = outputDir </> joinPath (map T.unpack (unModuleName (moduleName mod))) FP.<.> "ts"
-            createDirectoryIfMissing True (takeDirectory outputFile)
-            T.writeFileUtf8 outputFile modTxt
+      Right pm -> do
+        dependencies <-
+          forM (Map.toList pm) $
+            \(pkgId, (mbPkgName, pkg)) -> do
+                 let id = T.unpack $ unPackageId pkgId
+                     name = packageNameStr pkgId mbPkgName
+                     asName = if name == id then "itself" else name
+                 putStrLn $ "Generating " <> id <> " as " <> asName
+                 deps <- daml2ts opts pm pkgId pkg mbPkgName damlTypesVersion packageVersion
+                 pure (name, name, deps)
+        whenJust optInputPackageJson $
+          writeTopLevelPackageJson optOutputDir dependencies
 
-dup :: a -> (a, a)
-dup x = (x, x)
+packageNameStr :: PackageId -> Maybe PackageName -> String
+packageNameStr pkgId mbPkgIdent =
+  T.unpack (maybe (unPackageId pkgId) unPackageName mbPkgIdent)
 
-infixr 6 <.> -- This is the same fixity as '<>'.
-(<.>) :: T.Text -> T.Text -> T.Text
-(<.>) u v = u <> "." <> v
+newtype Scope = Scope {unscope :: String}
+newtype Dependency = Dependency {undependency :: String}  deriving (Eq, Ord)
 
-genModule :: Map.Map PackageId (Maybe PackageName, Package) -> PackageId -> Module -> Maybe T.Text
-genModule pm curPkgId mod
-  | null serDefs = Nothing
+-- Gives the scope '@foo' given a directory path like '/path/to/foo'.
+scopeOfScopeDir :: FilePath -> Scope
```

Comment:
We should make the scope configurable via a command line argument rather than infer it from some path. I suggest we use `@daml.js` as the default for that argument.

Reply:
Ok.

Reply:
https://github.com/digital-asset/daml/pull/4908